runners/direct-java/src/main/java/org/apache/beam/runners/direct/DoFnLifecycleManager.java

```diff
@@ -35,68 +37,68 @@
  * Manages {@link DoFn} setup, teardown, and serialization.
  *
  * <p>{@link DoFnLifecycleManager} is similar to a {@link ThreadLocal} storing a {@link DoFn}, but
- * calls the {@link DoFn} {@link Setup} the first time the {@link DoFn} is obtained and {@link
- * Teardown} whenever the {@link DoFn} is removed, and provides a method for clearing all cached
- * {@link DoFn DoFns}.
+ * calls the {@link DoFn} {@link Setup @Setup} method the first time the {@link DoFn} is obtained
+ * and {@link Teardown @Teardown} whenever the {@link DoFn} is removed, and provides a method for
+ * clearing all cached {@link DoFn DoFns}.
  */
 class DoFnLifecycleManager {
   private static final Logger LOG = LoggerFactory.getLogger(DoFnLifecycleManager.class);
 
-  public static DoFnLifecycleManager of(OldDoFn<?, ?> original) {
+  public static DoFnLifecycleManager of(DoFn<?, ?> original) {
     return new DoFnLifecycleManager(original);
   }
 
-  private final LoadingCache<Thread, OldDoFn<?, ?>> outstanding;
+  private final LoadingCache<Thread, DoFn<?, ?>> outstanding;
 
-  private DoFnLifecycleManager(OldDoFn<?, ?> original) {
+  private DoFnLifecycleManager(DoFn<?, ?> original) {
     this.outstanding = CacheBuilder.newBuilder().build(new DeserializingCacheLoader(original));
   }
 
-  public OldDoFn<?, ?> get() throws Exception {
+  public DoFn<?, ?> get() throws Exception {
     Thread currentThread = Thread.currentThread();
     return outstanding.get(currentThread);
   }
 
   public void remove() throws Exception {
     Thread currentThread = Thread.currentThread();
-    OldDoFn<?, ?> fn = outstanding.asMap().remove(currentThread);
-    fn.teardown();
+    DoFn<?, ?> fn = outstanding.asMap().remove(currentThread);
+    DoFnInvokers.INSTANCE.invokerFor(fn).invokeTeardown();
   }
 
   /**
    * Remove all {@link DoFn DoFns} from this {@link DoFnLifecycleManager}. Returns all exceptions
    * that were thrown while calling the remove methods.
    *
-   * <p>If the returned Collection is nonempty, an exception was thrown from at least one
-   * {@link DoFn#teardown()} method, and the {@link PipelineRunner} should throw an exception.
+   * <p>If the returned Collection is nonempty, an exception was thrown from at least one {@link
+   * DoFn.Teardown @Teardown} method, and the {@link PipelineRunner} should throw an exception.
    */
   public Collection<Exception> removeAll() throws Exception {
-    Iterator<OldDoFn<?, ?>> fns = outstanding.asMap().values().iterator();
+    Iterator<DoFn<?, ?>> fns = outstanding.asMap().values().iterator();
     Collection<Exception> thrown = new ArrayList<>();
     while (fns.hasNext()) {
-      OldDoFn<?, ?> fn = fns.next();
+      DoFn<?, ?> fn = fns.next();
       fns.remove();
       try {
-        fn.teardown();
+        DoFnInvokers.INSTANCE.invokerFor(fn).invokeTeardown();
```

Comment:
This should hold on to the invoker, rather than recreating it for each invocation.